pctasks/run/pctasks/run/task/batch.py

```diff
@@ -213,60 +240,121 @@ def get_failed_tasks(
             ],
             lambda x: x[0].batch_job_id,
         ):
-            indexed_task_ids: Dict[str, Tuple[str, str]] = {
+            indexed_task_ids = {
                 batch_id.batch_task_id: (partition_id, task_id)
                 for batch_id, (partition_id, task_id) in task_ids
             }
-            with BatchClient(self.settings.batch_settings) as batch_client:
-                for batch_task_id, error_message in batch_client.get_failed_tasks(
-                    job_id
-                ).items():
+
+            failed_tasks, job_failed = self._get_failed_tasks(job_id)
+            if job_failed:
+                for batch_task_id, (partition_id, task_id) in indexed_task_ids.items():
+                    if batch_task_id in failed_tasks:
+                        result[partition_id][task_id] = failed_tasks[batch_task_id]
+                    else:
+                        result[partition_id][task_id] = "Job failed"
+            else:
+                for batch_task_id, error_message in failed_tasks.items():
                     if batch_task_id in indexed_task_ids:
                         partition_id, task_id = indexed_task_ids[batch_task_id]
                         result[partition_id][task_id] = error_message
 
         return result
 
+    def _get_failed_tasks(self, job_id: str) -> Tuple[Dict[str, str], bool]:
+        """Checks for failed tasks in a Job.
+
+        Returns a dictionary of task IDs to error messages for specific task failures.
+        Returns True in the second value if the job failed.
+        """
+        # Check for cached task statuses
+        with failed_task_lock:
```

Comment:
Thinking out loud here, but I wonder if we can skip this lock, or at least minimize how long we lock for. I'm just on the lookout for deadlocks.

IIUC, we're caching responses by Job ID. What's the motivation for the lock?  Is it

1. To avoid writing "stale" information to the cache from two near-simultaneous HTTP requests, or
2. To avoid making the near-simultaneous HTTP requests in the first place?


If it's the first one, I would recommend moving the lock so that only the updates to the cache are locked. Then we don't have to worry about hangs in the HTTP request holding everything up.

If it's the second one, then the lock is probably OK. Though would it make sense to have a per-job lock, rather than a single lock? I'm not sure how often a single workflow has multiple jobs running. And I'm not sure how worried we are about reducing the number of concurrent calls to Batch (I imagine that whether the lock is global or per-job, we'll be making about the same number of concurrent requests since there's not often multiple jobs running concurrently, if ever?).

Reply:
Basically: For no particular reason I'm worried about deadlocks today :) We aren't going to be deadlocking with "ourselves", but if any operation inside this lock hangs then we'll pause all process on the orchestrator until the lock is released. So we should be pretty confident nothing in here is going to hang.

Reply:
There's a one-to-many relationship with Job ID to partition. So each partition will be asking for the failed task list, mostly from the same Job ID. This lock ensures that only one of those requests results in a call to the Batch API. Once that request is made from the first instance of the call, it is cached and the lock is released. The other threads waiting on that lock would then check the cache, find the result of the first call, and then move on without hitting the Batch API.

We could make the lock per-job. Currently, if there are two concurrent calls to this method for two different jobs, the results from the one job request would be blocked by the lock, and once unlocked, would potentially make another, separate API call because the results wouldn't necessarily be cached. With the current setup, the lock reduces the number of concurrent calls to the Batch API; with the per-job lock, we'll potentially be making a low number of concurrent calls to the Batch API, and avoid sequencing those (which will be a bit faster) while putting a bit more pressure on the API. That would also be more complicated lock management. 

Do you think that's worth it? If so I'll put it in.

As for deadlocks - I'm not sure where that would come in, unless the Batch API never returns. We've seen that it times out, so at worst the threads are waiting for the single Batch API call to return before it times out; once it times out, the cached result is an empty set so the threads wouldn't try the Batch API call again for at least 5 seconds. This improves current behavior, in which all threads would be hanging on concurrent Batch API calls and experiencing timeouts on each thread.

Reply:
> unless the Batch API never returns.

Right. Just thinking of the hangs in `ssl.py` that's been affecting goes-glm recently. But if that happens we're in trouble already, regardless of whether we have one thread hanging or all of them.

So overall, I just wanted to confirm that we're locking in the right spot. And it sounds like we are, based on the desire to reduce the number of concurrent API calls and keep things simple.